src/librender/scene.cpp

```diff
@@ -172,43 +172,117 @@ Scene<Float, Spectrum>::ray_test(const Ray3f &ray, Mask active) const {
         return ray_test_cpu(ray, active);
 }
 
+MTS_VARIANT std::tuple<typename Scene<Float, Spectrum>::Ray3f, Spectrum,
+                       const typename Scene<Float, Spectrum>::EmitterPtr>
+Scene<Float, Spectrum>::sample_emitter_ray(Float time, Float sample1,
+                                           const Point2f &sample2,
+                                           const Point2f &sample3,
+                                           Mask active) const {
+    MTS_MASKED_FUNCTION(ProfilerPhase::SampleEmitterRay, active);
+    using EmitterPtr = ek::replace_scalar_t<Float, Emitter*>;
+
+    if (unlikely(m_emitters.empty()))
+        return std::make_tuple(Ray3f(), Spectrum(0.f), EmitterPtr(nullptr));
+
+    // Randomly pick an emitter according to the precomputed emitter distribution
+    auto [index, emitter_weight, sample_re] = sample_emitter_reuse(sample1, active);
+    sample1 = sample_re;
+    EmitterPtr emitter = ek::gather<EmitterPtr>(m_emitters.data(), index, active);
+
+    // Wavelengths weight includes emitted radiance.
+    // TODO: update if wavelength sampling is changed
+#if 0
+    auto [wavelengths, wav_weight] =
+        emitter->sample_wavelengths(sample1, active);
+    auto [ray, ray_weight] =
+        emitter->sample_ray(time, wavelengths, sample2, sample3, active);
+    // Account for the discrete probability of sampling this emitter.
+    ray_weight *= wav_weight * emitter_weight;
+#else
+    auto [ray, ray_weight] =
+        emitter->sample_ray(time, sample1, sample2, sample3, active);
+#endif
+
+    return std::make_tuple(ray, ray_weight, emitter);
+}
+
+MTS_VARIANT std::pair<typename Scene<Float, Spectrum>::UInt32, Float>
```

Comment:
Can we only have one of `sample_emitter` and `sample_emitter_reuse`? (I would just keep the `_reuse` variant and then remove that suffix from the name)